File: algorithm/src/main/java/baekjoon/p01927/Main.java
package baekjoon.p01927;

import java.util.ArrayList;
import java.util.List;
import java.util.PriorityQueue;
import java.util.Scanner;

public class Main {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int[] inputs = new int[n];

        for (int i = 0; i < inputs.length; i++) {
            inputs[i] = sc.nextInt();
        }

        for (int solve : solve(inputs)) {
            System.out.println(solve);
        }
    }

    private static List<Integer> solve(int[] inputs) {
        List<Integer> ret = new ArrayList<>();
        MinHeap heap = new MinHeap();

        for (int input : inputs) {
            if (input == 0) {
                if (heap.isEmpty()) {
                    ret.add(0);
                } else {
                    ret.add(heap.remove());
                }
            } else {
                heap.add(input);
            }
        }

        return ret;
    }

    private static class MinHeap {

        private PriorityQueue<Integer> pq = new PriorityQueue<>();

        void add(int n) {
            pq.add(n);
        }

        int remove() {
            return pq.remove();
        }

        boolean isEmpty() {
            return pq.isEmpty();
        }

    }

}

File: algorithm/src/main/java/baekjoon/p01927/Main.kt
package baekjoon.p01927

import java.util.*

fun main(args: Array<String>) {
    val n = readLine()!!.toInt()
    val inputs = List(n) { readLine()!!.toInt() }
    println(solve(inputs).joinToString("\n"))
}

private fun solve(inputs: List<Int>): List<Int> {
    val ret = ArrayList<Int>()
    val heap = MinHeap()

    for (input in inputs) {
        when (input) {
            0 -> {
                when (heap.isEmpty()) {
                    true -> ret.add(0)
                    false -> ret.add(heap.remove())
                }
            }
            else -> {
                heap.add(input)
            }
        }
    }

    return ret
}

private class MinHeap {

    private val pq = PriorityQueue<Int>()

    fun isEmpty(): Boolean {
        return pq.isEmpty()
    }

    fun add(n: Int) {
        pq.add(n)
    }

    fun remove(): Int {
        return pq.remove()
    }

}